File: svc-data/src/test/java/com/beancounter/marketdata/TestInboundSerialization.java
package com.beancounter.marketdata;

import static org.assertj.core.api.Assertions.assertThat;

import com.beancounter.common.input.TrustedTrnEvent;
import com.beancounter.common.input.TrustedTrnImportRequest;
import com.beancounter.common.utils.BcJson;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.HashMap;
import org.junit.jupiter.api.Test;
import org.springframework.core.io.ClassPathResource;

public class TestInboundSerialization {
  private final ObjectMapper objectMapper = new BcJson().getObjectMapper();

  @Test
  void is_InboundPayloadConverted() throws Exception {
    TrustedTrnImportRequest payload = objectMapper.readValue(
        new ClassPathResource("/kafka/bc-view-send.json").getFile(),
        TrustedTrnImportRequest.class);

    assertThat(payload).isNotNull();

    HashMap<String, Object> message = objectMapper.readValue(
        new ClassPathResource("/kafka/csv-import-message.json").getFile(),
        new TypeReference<>() {
        }
    );

    assertThat(message).hasFieldOrProperty("payload");
    TrustedTrnImportRequest fromMsg = objectMapper.readValue(
        message.get("payload").toString(),
        TrustedTrnImportRequest.class);

    assertThat(fromMsg.getPortfolio()).usingRecursiveComparison().isEqualTo(payload.getPortfolio());
    assertThat(fromMsg.getRow()).contains(payload.getRow().toArray(new String[] {}));
    assertThat(fromMsg.getCallerRef()).isNull();
  }

  @Test
  void is_InboundMessagePayloadConverted() throws Exception {
    TrustedTrnImportRequest payload = objectMapper.readValue(
        new ClassPathResource("/kafka/bc-view-message.json").getFile(),
        TrustedTrnImportRequest.class);
    assertThat(payload).isNotNull();
  }

  @Test
  void is_IncomingTrustedEvent() throws Exception {
    TrustedTrnEvent inbound = objectMapper.readValue(
        new ClassPathResource("/kafka/event-incoming.json").getFile(), TrustedTrnEvent.class);
    assertThat(inbound).isNotNull();
  }

}

File: svc-data/src/test/java/com/beancounter/marketdata/TestInboundSerialization.kt
package com.beancounter.marketdata

import com.beancounter.common.input.TrustedTrnEvent
import com.beancounter.common.input.TrustedTrnImportRequest
import com.beancounter.common.utils.BcJson
import com.fasterxml.jackson.core.type.TypeReference
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Test
import org.springframework.core.io.ClassPathResource
import java.util.HashMap

class TestInboundSerialization {
    private val objectMapper = BcJson().objectMapper
    @Test
    @Throws(Exception::class)
    fun is_InboundPayloadConverted() {
        val payload = objectMapper.readValue(
            ClassPathResource("/kafka/bc-view-send.json").file,
            TrustedTrnImportRequest::class.java
        )
        Assertions.assertThat(payload).isNotNull
        val message: HashMap<String, Any> = objectMapper.readValue(
            ClassPathResource("/kafka/csv-import-message.json").file,
            object : TypeReference<HashMap<String, Any>>() {}
        )
        Assertions.assertThat(message).hasFieldOrProperty("payload")
        val (portfolio, _, callerRef, _, row) = objectMapper.readValue(
            message["payload"].toString(),
            TrustedTrnImportRequest::class.java
        )
        Assertions.assertThat(portfolio).usingRecursiveComparison().isEqualTo(payload.portfolio)
        Assertions.assertThat(row).contains(*payload.row.toTypedArray())
        Assertions.assertThat(callerRef).isNotNull
    }

    @Test
    @Throws(Exception::class)
    fun is_InboundMessagePayloadConverted() {
        val payload = objectMapper.readValue(
            ClassPathResource("/kafka/bc-view-message.json").file,
            TrustedTrnImportRequest::class.java
        )
        Assertions.assertThat(payload).isNotNull
    }

    @Test
    @Throws(Exception::class)
    fun is_IncomingTrustedEvent() {
        val inbound = objectMapper.readValue(
            ClassPathResource("/kafka/event-incoming.json").file, TrustedTrnEvent::class.java
        )
        Assertions.assertThat(inbound).isNotNull
    }
}
